File: src/phab_feedback/transport.py
"""Minimal HTTP transport with a mockable boundary."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Mapping, Protocol
from urllib.error import HTTPError, URLError
from urllib.parse import urlsplit
from urllib.request import Request, urlopen

from .errors import NetworkError


@dataclass(frozen=True)
class HttpResponse:
    status: int
    body: bytes
    headers: Mapping[str, str]


class Transport(Protocol):
    def request(
        self,
        method: str,
        url: str,
        *,
        headers: Mapping[str, str] | None = None,
        data: bytes | None = None,
    ) -> HttpResponse: ...


class UrllibTransport:
    def request(
        self,
        method: str,
        url: str,
        *,
        headers: Mapping[str, str] | None = None,
        data: bytes | None = None,
    ) -> HttpResponse:
        request = Request(url, data=data, method=method, headers=dict(headers or {}))
        try:
            with urlopen(request) as response:
                return HttpResponse(
                    status=response.status,
                    body=response.read(),
                    headers=dict(response.headers.items()),
                )
        except HTTPError as error:
            raise NetworkError(
                f"HTTP {error.code} from {_safe_location(url)}"
            ) from error
        except URLError as error:
            reason = getattr(error, "reason", "connection failed")
            raise NetworkError(
                f"Request to {_safe_location(url)} failed: {reason}"
            ) from error


def _safe_location(url: str) -> str:
    parsed = urlsplit(url)
    return f"{parsed.scheme}://{parsed.netloc}{parsed.path}"

File: internal/phabfeedback/transport.go
package phabfeedback

import (
	"fmt"
	"io"
	"net/http"
	"net/url"
)

type transport interface {
	Request(method, target string, headers http.Header, data io.Reader) ([]byte, error)
}

type httpTransport struct {
	client *http.Client
}

func (t httpTransport) Request(method, target string, headers http.Header, data io.Reader) ([]byte, error) {
	request, err := http.NewRequest(method, target, data)
	if err != nil {
		return nil, fmt.Errorf("create request: %w", err)
	}
	request.Header = headers.Clone()
	response, err := t.client.Do(request)
	if err != nil {
		return nil, newNetworkError(0, "Request to %s failed: %v", safeLocation(target), err)
	}
	defer response.Body.Close()
	body, err := io.ReadAll(response.Body)
	if err != nil {
		return nil, newNetworkError(response.StatusCode, "Could not read response from %s", safeLocation(target))
	}
	if response.StatusCode < 200 || response.StatusCode >= 300 {
		return nil, newNetworkError(response.StatusCode, "HTTP %d from %s", response.StatusCode, safeLocation(target))
	}
	return body, nil
}

func safeLocation(target string) string {
	parsed, err := url.Parse(target)
	if err != nil {
		return target
	}
	return parsed.Scheme + "://" + parsed.Host + parsed.Path
}
